pkg/reconciler/apis/apibindingdeletion/apibinding_deletor.go

```diff
@@ -0,0 +1,210 @@
+		)
+
+		return utilerrors.NewAggregate(deleteContentErrs)
+	}
+
+	if len(finalizersToNumRemaining) != 0 {
+		remainingByFinalizer := []string{}
+		for finalizer, numRemaining := range finalizersToNumRemaining {
+			if numRemaining == 0 {
+				continue
+			}
+			remainingByFinalizer = append(remainingByFinalizer, fmt.Sprintf("%s in %d resource instances", finalizer, numRemaining))
+		}
+		// sort for stable updates
+		sort.Strings(remainingByFinalizer)
+		conditions.MarkFalse(
+			apibinding,
+			apisv1alpha1.BindingResourceDeleteSuccess,
+			ResourceFinalizersRemainReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			fmt.Sprintf("Some content in the workspace has finalizers remaining: %s", strings.Join(remainingByFinalizer, ", ")),
+		)
+
+		return &deletion.ResourcesRemainingError{Estimate: DeletionRecheckEstimateSeconds}
+	}
+
+	if len(gvrToNumRemaining) != 0 {
+		remainingResources := []string{}
+		for gvr, numRemaining := range gvrToNumRemaining {
+			if numRemaining == 0 {
+				continue
+			}
+			remainingResources = append(remainingResources, fmt.Sprintf("%s.%s has %d resource instances", gvr.Resource, gvr.Group, numRemaining))
+		}
+		// sort for stable updates
+		sort.Strings(remainingResources)
+
+		conditions.MarkFalse(
+			apibinding,
+			apisv1alpha1.BindingResourceDeleteSuccess,
+			ResourceRemainingReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			fmt.Sprintf("Some resources are remaining: %s", strings.Join(remainingResources, ", ")),
+		)
+
+		return &deletion.ResourcesRemainingError{Estimate: DeletionRecheckEstimateSeconds}
+	}
+
+	conditions.MarkTrue(apibinding, apisv1alpha1.BindingResourceDeleteSuccess)
+
+	return nil
+}
+
+func (c *Controller) deleteAllCR(ctx context.Context, clusterName logicalcluster.Name, gvr schema.GroupVersionResource) (gvrDeletionMetadata, error) {
+	partialList, err := c.metadataClient.Resource(gvr).Namespace(metav1.NamespaceAll).List(genericapirequest.WithCluster(ctx, genericapirequest.Cluster{Name: clusterName}), metav1.ListOptions{})
+	if err != nil {
+		return gvrDeletionMetadata{}, err
+	}
+
+	deletedNamespaces := sets.String{}
+	deleteErrors := []error{}
+
+	for _, item := range partialList.Items {
+		if deletedNamespaces.Has(item.GetNamespace()) {
+			continue
+		}
+
+		// don't retry deleting the same namespace
+		deletedNamespaces.Insert(item.GetNamespace())
+		background := metav1.DeletePropagationBackground
+		opts := metav1.DeleteOptions{PropagationPolicy: &background}
+
+		// CRs always support deletecollection verb
+		if err := c.metadataClient.Resource(gvr).Namespace(item.GetNamespace()).DeleteCollection(
+			genericapirequest.WithCluster(ctx, genericapirequest.Cluster{Name: clusterName}), opts, metav1.ListOptions{}); err != nil {
+			deleteErrors = append(deleteErrors, err)
+			continue
+		}
+	}
+
+	deleteError := utilerrors.NewAggregate(deleteErrors)
+	if deleteError != nil {
+		return gvrDeletionMetadata{}, err
+	}
+
+	// list again so we know how many left resources
```

Comment:
This seems brittle, we just hope that during the time taken to issue the background delete calls some have finished?